File: 2020/7.py
#!/usr/bin/env python
# Day 7

class Bag(object):
  def __init__(self, name):
    self.name = name
    self.contains = {} # {str: count} # TODO: change to Bag?

  def __repr__(self):
    return "Bag(%s -> %s)" % (self.name, self.contains.keys())

inputfile = "input7.txt"

lines = [
  "light red bags contain 1 bright white bag, 2 muted yellow bags",
  "dark orange bags contain 3 bright white bags, 4 muted yellow bags",
  "bright white bags contain 1 shiny gold bag",
  "muted yellow bags contain 2 shiny gold bags, 9 faded blue bags",
  "shiny gold bags contain 1 dark olive bag, 2 vibrant plum bags",
  "dark olive bags contain 3 faded blue bags, 4 dotted black bags",
  "vibrant plum bags contain 5 faded blue bags, 6 dotted black bags",
  "faded blue bags contain no other bags",
  "dotted black bags contain no other bags",
]
lines = [
  "shiny gold bags contain 2 dark red bags",
  "dark red bags contain 2 dark orange bags",
  "dark orange bags contain 2 dark yellow bags",
  "dark yellow bags contain 2 dark green bags",
  "dark green bags contain 2 dark blue bags",
  "dark blue bags contain 2 dark violet bags",
  "dark violet bags contain no other bags",
]
with open(inputfile, "r") as f:
  lines = [x.strip().strip(".") for x in f.readlines()]

bags = {}  # {str: Bag, ...}
for line in lines:
  outer_name, contained = line.replace("bags", "").replace("bag", "").strip().split(" contain ")
  outer_name = outer_name.strip()
  if outer_name in bags:
    outer_bag = bags[outer_name]
  else:
    outer_bag = Bag(outer_name)
    bags[outer_name] = outer_bag

  inners = [x.strip() for x in contained.split(",")]
  for inner in inners:
    if inner == "no other":
      continue
    first = inner.split(" ")[0].strip()
    bag_name = inner[len(first):].strip()
    if bag_name not in bags:
      bags[bag_name] = Bag(bag_name)
    count = int(first)
    outer_bag.contains[bag_name] = count

# Part one
seen = set()
can_reach_gold = set(["shiny gold"])

def look_inside(outer): # Bag
  if outer.name in can_reach_gold:
    return True
  for bag_name in outer.contains.keys():
    bag = bags[bag_name]
    if look_inside(bag):
      can_reach_gold.add(outer.name)
      return True
  return False


for bag in bags.values():
  look_inside(bag)

print can_reach_gold
print len(can_reach_gold) - 1

# Part two
def count_inside(outer):
  if not outer.contains:
    return 0
  count = 0
  for bag, number in outer.contains.iteritems():
    count += number
    count += (number * count_inside(bags[bag]))
  return count

current = bags["shiny gold"]
print count_inside(current)

File: 2020/7.kt
// Advent of code 2020 day 7
import java.io.File

class Bag(name: String) {
    var name = name
    var contains = mutableMapOf<String, Int>() // {str: count}
}

fun main() {
  val fileName = "input7.txt"
  val lines: List<String> = File(fileName).readLines()

  var bags = mutableMapOf<String, Bag>()
  for (line in lines) {
    val pieces = line.replace("bags", "").replace("bag", "").replace(".", "").trim().split(" contain ")
    val outer_name = pieces[0].trim()
    val contained = pieces[1]
    var outer_bag: Bag
    if (bags.containsKey(outer_name)) {
      outer_bag = bags[outer_name]!!
    } else {
      outer_bag = Bag(outer_name)
      bags[outer_name] = outer_bag
    }

    val inners = contained.split(",").map { it.trim() }
    for (inside in inners) {
      if (inside == "no other") {
        continue
      }
      val first:String = inside.split(" ")[0].trim()
      val bag_name:String = inside.substring(first.length).trim()
      if (!bags.containsKey(bag_name)) {
        bags[bag_name] = Bag(bag_name)
      }
      outer_bag.contains[bag_name] = first.toInt()
    }
  }
  // Part one
  var can_reach_gold = mutableSetOf("shiny gold")

  for (bag in bags.values) {
    look_inside(bag, bags, can_reach_gold)
  }

  // Part 1
  println("Part 1: ${can_reach_gold.size - 1}")

  // Part two
  val bag_count = count_inside(bags["shiny gold"]!!, bags)
  println("Part 2: $bag_count bags")
}

fun count_inside(outer: Bag, bags: Map<String, Bag>): Int {
  if (outer.contains.size == 0) {
    return 0
  }
  var count = 0
  for ((bag, number) in outer.contains) {
    count += number
    count += (number * count_inside(bags[bag]!!, bags))
  }
  return count
}


fun look_inside(outer: Bag, bags: Map<String, Bag>, can_reach_gold: MutableSet<String>): Boolean {
  if (outer.name in can_reach_gold) {
    return true
  }
  for (bag_name in outer.contains.keys) {
    val bag = bags[bag_name]!!
    if (look_inside(bag, bags, can_reach_gold)) {
      can_reach_gold.add(outer.name)
      return true
    }
  }
  return false
}


